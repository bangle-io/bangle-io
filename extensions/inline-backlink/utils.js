export function removeMdExtension(str) {
  if (str.endsWith('.md')) {
    return str.slice(0, -3);
  }
  return str;
}
