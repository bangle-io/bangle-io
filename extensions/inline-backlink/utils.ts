import {
  filePathToWsPath,
  removeExtension,
  resolvePath,
  sanitizeFilePath,
  suffixWithNoteExtension,
  validateWsPath,
} from '@bangle.io/ws-path';

export function wsPathFromQuery(query: string, wsName: string) {
  let filePath = query.split(']]').join('');
  filePath = filePath.trim();
  filePath = suffixWithNoteExtension(filePath);
  let wsPath = filePathToWsPath(wsName, filePath);
  try {
    validateWsPath(wsPath);
  } catch (error) {
    wsPath = filePathToWsPath(wsName, sanitizeFilePath(filePath));
  }

  return wsPath;
}

export function getBacklinkPath(wsPath: string, allWsPaths: readonly string[]) {
  const { fileName, filePath } = resolvePath(wsPath);
  const matchingFilenames = allWsPaths.filter(
    (w) => resolvePath(w).fileName === fileName,
  );

  if (matchingFilenames.length === 0) {
    return removeExtension(filePath);
  }

  // if there are multiple files with the same name
  // give it an absolute path
  if (matchingFilenames.length > 1) {
    return removeExtension(filePath);
  }

  return removeExtension(fileName);
}
