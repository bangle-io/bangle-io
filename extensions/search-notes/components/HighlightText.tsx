import React from 'react';

import type { HighlightTextType } from '@bangle.io/search-pm-node';

export function HighlightText({
  highlightText,
}: {
  highlightText: HighlightTextType;
}) {
  return (
    <div className="highlight-text-container">
      {highlightText.map((t, i) => (
        <span key={i} className="highlight-text text-sm">
          {t}
        </span>
      ))}
    </div>
  );
}
