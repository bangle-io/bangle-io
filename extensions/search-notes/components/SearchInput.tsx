import React, { useCallback, useEffect, useRef } from 'react';

import { Input } from '@bangle.io/ui-components';

export const SearchInput = ({
  searchQuery,
  updateSearchQuery,
}: {
  searchQuery: string;
  updateSearchQuery: (q: string) => void;
}) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const onClear = useCallback(
    (e) => {
      e.preventDefault();
      updateSearchQuery('');
    },
    [updateSearchQuery],
  );

  const onChange = useCallback(
    (e) => {
      updateSearchQuery(e.target.value);
    },
    [updateSearchQuery],
  );

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <Input
      label="Search"
      ref={inputRef}
      className="w-full"
      showClear
      value={searchQuery}
      onClear={onClear}
      onChange={onChange}
      placeholder="Search"
    />
  );
};
