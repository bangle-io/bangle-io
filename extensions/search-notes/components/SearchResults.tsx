import React, { useCallback, useEffect, useMemo, useState } from 'react';

import { NodeSelection, Selection } from '@bangle.dev/pm';

import { nsmApi2, wsPathHelpers } from '@bangle.io/api';
import { NoteLink } from '@bangle.io/contextual-ui-components';
import type { SearchMatch, SearchResultItem } from '@bangle.io/search-pm-node';
import {
  ButtonIcon,
  ChevronDownIcon,
  ChevronRightIcon,
  Sidebar,
} from '@bangle.io/ui-components';
import { cx, safeRequestAnimationFrame, usePrevious } from '@bangle.io/utils';

import { HighlightText } from './HighlightText';

function useCollapseMarker(
  results: SearchResultItem[],
  // collapses all results if counter is greater than zero integer
  // increment its value to retrigger the collapse
  collapseAllCounter: number,
) {
  const [cache, updateCache] = useState(new Set());

  const onClicks = useMemo(
    () =>
      results.map((r) => {
        const onClick = () => {
          updateCache((c) => {
            const s = new Set(c);

            if (s.has(r)) {
              s.delete(r);
            } else {
              s.add(r);
            }

            return s;
          });
        };

        return onClick;
      }),
    [results],
  );
  const isCollapsed = useCallback(
    (r) => {
      return cache.has(r);
    },
    [cache],
  );

  useEffect(() => {
    if (collapseAllCounter > 0) {
      updateCache(() => {
        return new Set(results);
      });
    }
  }, [collapseAllCounter, results]);

  return { onClicks, isCollapsed };
}

export function SearchResults({
  results,
  collapseAllCounter,
}: {
  collapseAllCounter: number;
  results: SearchResultItem[];
}) {
  const { onClicks, isCollapsed } = useCollapseMarker(
    results,
    collapseAllCounter,
  );

  const { widescreen } = nsmApi2.ui.useUi();
  const { primaryWsPath } = nsmApi2.workspace.useWorkspace();

  const [currentlyClicked, updateCurrentlyClicked] = useState<null | {
    wsPath: string;
    match: SearchMatch;
    matchIndex: number;
  }>();

  useEffect(() => {
    let cancelled = false;

    if (currentlyClicked && primaryWsPath === currentlyClicked.wsPath) {
      // TODO this is a mess, we need a better api to know when the editor is ready
      setTimeout(() => {
        safeRequestAnimationFrame(() => {
          const editor = nsmApi2.editor.getPrimaryEditor();

          if (cancelled || !editor || editor.destroyed) {
            return;
          }
          editor.focusView();
          const { dispatch, state } = editor.view;
          let tr = state.tr;

          if (currentlyClicked.match.parentPos >= tr.doc.content.size) {
            tr = tr.setSelection(Selection.atEnd(tr.doc));
          } else {
            // The following is a hack which solves the purpose
            // to grab the users attention. It is not great but
            // it was easy to implement. In future we can move to a
            // a more formal less hacky way to grab attention that
            // doesn't involve messing with selections.
            // its Working :-
            //  - show a node selection .. ends up drawing a big rectangle
            //    outline on parent node
            //  - after some time set a text selection clearing the rectangle

            const parentNode = tr.doc.nodeAt(currentlyClicked.match.parentPos);

            if (parentNode) {
              tr = tr.setSelection(
                NodeSelection.create(tr.doc, currentlyClicked.match.parentPos),
              );
            }
            setTimeout(() => {
              if (!editor.destroyed) {
                const { dispatch, state } = editor.view;
                const tr = state.tr;
                dispatch(
                  tr.setSelection(
                    Selection.near(
                      tr.doc.resolve(currentlyClicked.match.parentPos),
                    ),
                  ),
                );
              }
            }, 300);
          }
          try {
            dispatch(tr.scrollIntoView());
          } catch (error) {
            // ignore because of a bug where a user edits a note and clicks on the already searched
            // resulting in dom error
          }

          updateCurrentlyClicked(null);
        });
      }, 50);
    }

    return () => {
      cancelled = true;
    };
  }, [currentlyClicked, primaryWsPath]);

  // cannot used currently clicked because it gets set to null
  // right after focusing
  const prevClicked = usePrevious(currentlyClicked);

  return (
    <>
      {results.map((r, i) => {
        const wsPath = wsPathHelpers.createWsPath(r.uid);

        return (
          <React.Fragment key={i}>
            <Sidebar.Row2
              titleClassName="text-sm font-bold"
              className={cx(
                `B-search-notes_search-result-note-match pl-1 pr-3  select-none`,
                primaryWsPath === r.uid && 'BU_active',
              )}
              extraInfoClassName="text-sm"
              extraInfoOnNewLine
              onClick={onClicks[i]}
              item={{
                uid: 'search-notes-result-' + i,
                showDividerAbove: false,
                title: wsPathHelpers.resolvePath2(wsPath).fileNameWithoutExt,
                extraInfo: wsPathHelpers.resolvePath2(wsPath).dirPath,
                leftNode: (
                  <ButtonIcon>
                    {isCollapsed(r) ? (
                      <ChevronRightIcon style={IconStyle} />
                    ) : (
                      <ChevronDownIcon style={IconStyle} />
                    )}
                  </ButtonIcon>
                ),
                rightNode: (
                  <ButtonIcon className="text-xs font-semibold">
                    {r.matches.length}
                  </ButtonIcon>
                ),
                rightHoverNode: (
                  <ButtonIcon className="text-xs font-semibold">
                    {r.matches.length}
                  </ButtonIcon>
                ),
              }}
            />
            {!isCollapsed(r) &&
              r.matches.map((matchObj, j) => {
                return (
                  <NoteLink
                    wsPath={r.uid}
                    className={cx(
                      'rounded-sm block B-search-notes_search-result-text-match ',
                      primaryWsPath === r.uid &&
                        prevClicked?.matchIndex === j &&
                        'B-search-notes_previously-clicked',
                      j === 0 ? 'mt-3' : 'mt-4',
                      j === r.matches.length - 1
                        ? 'pb-4 border-b-1 border-colorNeutralBorder'
                        : '',
                    )}
                    onClick={() => {
                      const primaryEditor = nsmApi2.editor.getPrimaryEditor();

                      // if not in a widescreen close the sidebar
                      // after click
                      if (!widescreen) {
                        nsmApi2.ui.closeSidebar();
                      } else if (primaryEditor && !primaryEditor.destroyed) {
                        updateCurrentlyClicked({
                          wsPath: r.uid,
                          match: matchObj,
                          matchIndex: j,
                        });
                      }
                    }}
                    key={j}
                  >
                    <Sidebar.Row2
                      className="pl-3"
                      titleClassName="text-sm "
                      item={{
                        uid: 'search-result-text-match-' + j,
                        title: (
                          <HighlightText
                            highlightText={matchObj.match}
                          ></HighlightText>
                        ),
                      }}
                    ></Sidebar.Row2>
                  </NoteLink>
                );
              })}
          </React.Fragment>
        );
      })}
    </>
  );
}

const IconStyle = {
  height: 16,
  width: 16,
};
