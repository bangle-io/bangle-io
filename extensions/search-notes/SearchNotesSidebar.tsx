import { useActionContext } from 'action-context';
import React, { useContext, useEffect, useState } from 'react';
import { ButtonIcon, Sidebar, SpinnerIcon } from 'ui-components';
import { useWorkspaceContext } from 'workspace-context';
import { useSearchNotes } from './hooks';
import { SearchInput } from './SearchInput';
import { SearchResults } from './SearchResults';
import { search } from '@bangle.dev/search';
import { useEditorManagerContext } from 'editor-manager-context';
import { searchPluginKey } from './plugin-key';

export function SearchNotesSidebar() {
  const { dispatchAction } = useActionContext();
  const { forEachEditor } = useEditorManagerContext();
  const [query, updateQuery] = useState('');
  const { wsName } = useWorkspaceContext();
  const [collapseAllCounter, updateCollapseAllCounter] = useState(0);
  const { results, pendingSearch } = useSearchNotes(query);
  useEffect(() => {
    updateCollapseAllCounter(0);
  }, [query, wsName]);

  useEffect(() => {
    if (results && results.length > 0 && query) {
      forEachEditor((editor) => {
        if (editor?.destroyed === false) {
          const queryRegex = query
            ? new RegExp(query.replace(/[-[\]{}()*+?.,\\^$|]/g, '\\$&'))
            : undefined;

          search.updateSearchQuery(searchPluginKey, queryRegex)(
            editor.view.state,
            editor.view.dispatch,
          );
        }
      });
    }
  }, [forEachEditor, query, results]);

  useEffect(() => {
    return () => {
      forEachEditor((editor) => {
        if (editor?.destroyed === false) {
          search.updateSearchQuery(searchPluginKey, undefined)(
            editor.view.state,
            editor.view.dispatch,
          );
        }
      });
    };
  }, [forEachEditor]);

  if (!wsName) {
    return (
      <div className="h-full flex flex-col items-center justify-center">
        <span
          className="text-sm font-extrabold b-text-color-lighter cursor-pointer"
          onClick={() => {
            dispatchAction({
              name: '@action/core-palettes/TOGGLE_WORKSPACE_PALETTE',
            });
          }}
        >
          Please open a workspace to search
        </span>
      </div>
    );
  }

  return (
    <Sidebar.Container className="search-notes">
      <Sidebar.Title className="mt-2 px-2">üîç Search notes</Sidebar.Title>
      <Sidebar.ItemContainer className="mt-2 px-2">
        <SearchInput updateQuery={updateQuery} query={query} />
      </Sidebar.ItemContainer>
      <Sidebar.ItemContainer className="flex flex-row justify-between my-1 px-2 text-xs">
        {results && (
          <span className="">
            {results.length === 0
              ? 'No match found'
              : `Found ${results.length} notes`}
          </span>
        )}
        {results && (
          <ButtonIcon
            onClick={() => {
              updateCollapseAllCounter((c) => c + 1);
            }}
          >
            Collapse all
          </ButtonIcon>
        )}
        {pendingSearch && (
          <>
            <span className="">Searching </span>
            <span className="h-3 w-3 mr-2">
              <SpinnerIcon />
            </span>
          </>
        )}
      </Sidebar.ItemContainer>
      <Sidebar.ScrollableContainer>
        {results && (
          <SearchResults
            results={results}
            collapseAllCounter={collapseAllCounter}
          />
        )}
      </Sidebar.ScrollableContainer>
    </Sidebar.Container>
  );
}
