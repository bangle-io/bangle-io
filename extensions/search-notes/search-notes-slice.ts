import { nsm, nsmApi2 } from '@bangle.io/api';
import { isAbortError } from '@bangle.io/mini-js-utils';

import type { SearchNotesExtensionState } from './constants';
import { searchNotes } from './search-notes';

const initState: SearchNotesExtensionState = {
  searchQuery: '',
  externalInputChange: 0,
  pendingSearch: false,
  searchResults: null,
};

export const SLICE_NAME = 'slice::search-notes:main';

export const searchSlice = nsm.slice([], {
  name: SLICE_NAME,
  state: initState,
});

export const updateSearchQuery = searchSlice.action(
  function updateSearchQuery(data: { query: string }) {
    return searchSlice.tx((state) => {
      return searchSlice.update(state, {
        searchQuery: data.query,
      });
    });
  },
);

export const updateExternalSearchQuery = searchSlice.action(
  function updateExternalSearchQuery(query: string) {
    return searchSlice.tx((state) => {
      return searchSlice.update(state, (sliceState) => {
        return {
          searchQuery: query,
          externalInputChange: sliceState.externalInputChange + 1,
        };
      });
    });
  },
);

export const updateSliceState = searchSlice.action(function updateSliceState(
  data: Omit<Partial<SearchNotesExtensionState>, 'searchQuery'>,
) {
  return searchSlice.tx((state) => {
    return searchSlice.update(state, data);
  });
});

const wsNameEffect = nsm.effect(function wsNameEffect(store) {
  void nsmApi2.workspace.trackWorkspaceName(store);

  store.dispatch(
    updateSliceState({
      searchResults: null,
      pendingSearch: false,
    }),
  );
});

const searchEffect = nsm.effect(function searchEffect(store) {
  const wsName = nsmApi2.workspace.trackWorkspaceName(store);

  const { searchQuery } = searchSlice.track(store);

  if (!wsName) {
    return;
  }

  if (searchQuery === '') {
    store.dispatch(
      updateSliceState({ searchResults: null, pendingSearch: false }),
    );

    return;
  }

  store.dispatch(
    updateSliceState({ searchResults: null, pendingSearch: true }),
  );

  let controller = new AbortController();

  nsm.cleanup(store, () => {
    controller.abort();
  });

  searchNotes(controller.signal, searchQuery, wsName)
    .then((result) => {
      if (!controller.signal.aborted) {
        store.dispatch(
          updateSliceState({ searchResults: result, pendingSearch: false }),
        );
      } else {
        console.warn('Search aborted', result);
      }
    })
    .catch((error) => {
      if (isAbortError(error)) {
        return;
      }
      store.dispatch(updateSliceState({ pendingSearch: false }));

      throw error;
    });
});

const highlightEditorsEffect = nsm.effect(function highlightEditorsEffect(
  store,
) {
  const { searchQuery } = searchSlice.track(store);

  if (!searchQuery) {
    return;
  }

  const queryRegex = searchQuery
    ? new RegExp(searchQuery.replace(/[-[\]{}()*+?.,\\^$|]/g, '\\$&'), 'i')
    : undefined;

  nsmApi2.editor.updateEditorSearchQuery(queryRegex);
});

export const searchEffects = [
  wsNameEffect,
  searchEffect,
  highlightEditorsEffect,
];
