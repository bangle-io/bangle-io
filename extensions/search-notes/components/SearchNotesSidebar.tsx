import React, { useEffect, useState } from 'react';

import { nsmApi2, useNsmSlice } from '@bangle.io/api';
import { CorePalette } from '@bangle.io/constants';
import { ButtonIcon, Sidebar, SpinnerIcon } from '@bangle.io/ui-components';
import { useDebouncedValue } from '@bangle.io/utils';

import { searchSlice, updateSearchQuery } from '../search-notes-slice';
import { SearchInput } from './SearchInput';
import { SearchResults } from './SearchResults';

export function SearchNotesSidebar() {
  const { wsName } = nsmApi2.workspace.useWorkspace();

  const [
    {
      pendingSearch,
      searchResults,
      searchQuery,
      externalInputChange: externalChange,
    },
    searchDispatch,
  ] = useNsmSlice(searchSlice);

  const [collapseAllCounter, updateCollapseAllCounter] = useState(0);

  const [rawSearchQuery, updateRawSearchQuery] = useState(searchQuery || '');

  const [lastExternalChange, setLastExternalChange] = useState(externalChange);

  const localSearchQuery = useDebouncedValue(rawSearchQuery, {
    wait: 30,
  });

  useEffect(() => {
    // reset the current input state if there was an external
    // change made to the search query
    if (lastExternalChange !== externalChange) {
      setLastExternalChange(externalChange);
      updateRawSearchQuery(searchQuery);
    } else {
      if (localSearchQuery !== searchQuery) {
        // else sync the local state with the global state
        searchDispatch(updateSearchQuery({ query: localSearchQuery }));
      }
    }
  }, [
    externalChange,
    searchQuery,
    localSearchQuery,
    lastExternalChange,
    searchDispatch,
  ]);

  useEffect(() => {
    updateCollapseAllCounter(0);
  }, [localSearchQuery, wsName]);

  if (!wsName) {
    return (
      <div className="flex flex-col items-center justify-center h-full">
        <span
          className="text-sm font-extrabold cursor-pointer text-colorNeutralTextSubdued"
          onClick={() => {
            nsmApi2.ui.togglePalette(CorePalette.Workspace);
          }}
        >
          Please open a workspace to search
        </span>
      </div>
    );
  }

  return (
    <Sidebar.Container className="B-search-notes_search-notes">
      <Sidebar.ItemContainer className="px-2 mt-2">
        <SearchInput
          searchQuery={rawSearchQuery}
          updateSearchQuery={updateRawSearchQuery}
        />
      </Sidebar.ItemContainer>
      <Sidebar.ItemContainer className="flex flex-row justify-between px-2 my-1 text-xs">
        {searchResults && (
          <span className="">
            {searchResults.length === 0
              ? 'No match found'
              : `Found ${searchResults.length} notes`}
          </span>
        )}
        {searchResults && (
          <ButtonIcon
            onClick={() => {
              updateCollapseAllCounter((c) => c + 1);
            }}
          >
            Collapse all
          </ButtonIcon>
        )}
        {pendingSearch && (
          <>
            <span className="">Searching </span>
            <span className="w-3 h-3 mr-2">
              <SpinnerIcon />
            </span>
          </>
        )}
      </Sidebar.ItemContainer>
      <Sidebar.ScrollableContainer>
        {searchResults && (
          <SearchResults
            results={searchResults}
            collapseAllCounter={collapseAllCounter}
          />
        )}
      </Sidebar.ScrollableContainer>
    </Sidebar.Container>
  );
}
