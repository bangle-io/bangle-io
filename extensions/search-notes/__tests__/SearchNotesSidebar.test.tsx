/**
 * @jest-environment @bangle.io/jsdom-env
 */
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import React from 'react';
import searchNotesExtension from '../index';

import { setupTestExtension, waitForExpect } from '@bangle.io/test-utils-2';
import { SearchNotesSidebar } from '../components/SearchNotesSidebar';
import { nsmApi2, wsPathHelpers } from '@bangle.io/api';
import { sleep } from '@bangle.io/utils';
import { SearchResultItem } from '@bangle.io/search-pm-node';
import { searchSlice, updateSliceState } from '../search-notes-slice';
let abortController = new AbortController();

beforeEach(() => {
  abortController = new AbortController();
});

afterEach(async () => {
  abortController.abort();
});

const setup = async (
  wsName: string | null = 'test-ws',
  // Array of [wsPath, MarkdownString]
  noteWsPaths: Array<[string, string]> = [
    [`${wsName}:one.md`, `# Hello World 0`],
    [`${wsName}:two.md`, `# Hello World 1`],
  ],
) => {
  const ctx = setupTestExtension({
    abortSignal: abortController.signal,
    extensions: [searchNotesExtension],
    editor: true,
    storeName: 'store:' + wsName,
  });

  if (wsName) {
    await ctx.createWorkspace(wsName);
    await ctx.createNotes(noteWsPaths, { loadFirst: true });

    const targetPath = wsPathHelpers.createWsPath(noteWsPaths?.[0]?.[0]);
    nsmApi2.workspace.pushPrimaryWsPath(targetPath);

    await waitForExpect(() => {
      expect(nsmApi2.workspace.workspaceState().primaryWsPath).toEqual(
        targetPath,
      );
    });
  }

  return {
    store: ctx.testStore,
    render: async () => {
      let counter = 0;
      let { container, rerender } = render(
        <ctx.ContextProvider>
          <SearchNotesSidebar />
        </ctx.ContextProvider>,
      );

      return {
        container,
        rerender: () => {
          counter++;

          return rerender(
            <ctx.ContextProvider>
              <SearchNotesSidebar />
            </ctx.ContextProvider>,
          );
        },
      };
    },
  };
};

test('renders correctly is no workspace is opened', async () => {
  const { render } = await setup(null);
  const { container } = await render();

  expect(container).toMatchInlineSnapshot(`
    <div>
      <div
        class="flex flex-col items-center justify-center h-full"
      >
        <span
          class="text-sm font-extrabold cursor-pointer text-colorNeutralTextSubdued"
        >
          Please open a workspace to search
        </span>
      </div>
    </div>
  `);
});

test('renders correctly workspace is opened', async () => {
  const { render } = await setup('test-ws-2');
  const { container } = await render();

  expect(container).toMatchSnapshot();
});

test('show search results', async () => {
  const { render, store } = await setup('test-ws-3');

  const searchResultItem: SearchResultItem = {
    uid: 'test-ws-3:one.md',
    matches: [{ parent: 'para', parentPos: 0, match: ['hello', ' world'] }],
  };

  const { container, rerender } = await render();

  store.dispatch(
    updateSliceState({
      searchResults: [searchResultItem],
      pendingSearch: false,
    }),
  );

  rerender();

  expect(
    Array.from(
      container.querySelectorAll('.B-search-notes_search-result-note-match'),
    ).length,
  ).toBe(1);

  expect(
    Array.from(
      container.querySelectorAll('[data-id="search-result-text-match-0"]'),
    ).length,
  ).toBe(1);

  expect(container).toMatchSnapshot();
});

test('typing in input triggers extension state update', async () => {
  const { render, store } = await setup('test-ws-4');

  render();

  const input = screen.getByLabelText('Search', { selector: 'input' });

  fireEvent.change(input, { target: { value: 'hello' } });

  await waitForExpect(() => {
    expect(searchSlice.get(store.state).searchQuery).toEqual('hello');
  });
});

test('pendingSearch shows a spinner', async () => {
  const { render, store } = await setup('test-ws-5');

  const { rerender } = await render();

  store.dispatch(
    updateSliceState({
      pendingSearch: true,
    }),
  );

  rerender();

  expect(Boolean(screen.queryByLabelText('search spinner'))).toBe(true);

  store.dispatch(
    updateSliceState({
      pendingSearch: false,
    }),
  );
  rerender();

  expect(Boolean(screen.queryByLabelText('search spinner'))).toBe(false);
});

describe('search results', () => {
  test('works with search query', async () => {
    const { store, render } = await setup('test-ws-6', [
      [`test-ws-6:one.md`, `# Hello kjoshi 0`],
      [`test-ws-6:two.md`, `# Hello World 1`],
      [
        `test-ws-6:three.md`,
        `# Hello World 3 

>       kjoshi`,
      ],
    ]);

    const { rerender } = await render();

    const input = screen.getByLabelText('Search', { selector: 'input' });

    fireEvent.change(input, { target: { value: 'kjoshi' } });

    rerender();

    await waitFor(() => {
      expect(screen.getByText('Found 2 notes')).toBeDefined();
    });

    expect(screen.queryAllByText('kjoshi')).toMatchInlineSnapshot(`
      [
        <span
          class="B-search-notes_highlight-text text-sm"
        >
          kjoshi
        </span>,
        <span
          class="B-search-notes_highlight-text text-sm"
        >
          kjoshi
        </span>,
      ]
    `);

    // firing another query works
    fireEvent.change(input, { target: { value: 'hello' } });

    rerender();

    await waitFor(() => {
      expect(screen.getByText('Found 3 notes')).toBeDefined();
    });
  });
});
