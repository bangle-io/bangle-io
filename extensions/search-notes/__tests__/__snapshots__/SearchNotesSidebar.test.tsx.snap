// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly workspace is opened 1`] = `
<div>
  <div
    class="flex flex-col h-full B-search-notes_search-notes"
    style="overflow-y: auto;"
  >
    <div
      class="px-2 mt-2 "
    >
      <div
        class="B-ui-components_input-container"
      >
        <input
          aria-label="Search"
          autocapitalize="off"
          autocorrect="off"
          class="outline-offset-1 text-field-neutral border-1 border-colorNeutralTextFieldBorder rounded px-2 py-1 w-full  B-ui-components_misc-input-ring "
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          style="position: relative; display: flex;"
        />
      </div>
    </div>
    <div
      class="flex flex-row justify-between px-2 my-1 text-xs "
    />
    <div
      class="flex-grow"
      style="overflow-y: auto;"
    />
  </div>
</div>
`;

exports[`show search results 1`] = `
<div>
  <div
    class="flex flex-col h-full B-search-notes_search-notes"
    style="overflow-y: auto;"
  >
    <div
      class="px-2 mt-2 "
    >
      <div
        class="B-ui-components_input-container"
      >
        <input
          aria-label="Search"
          autocapitalize="off"
          autocorrect="off"
          class="outline-offset-1 text-field-neutral border-1 border-colorNeutralTextFieldBorder rounded px-2 py-1 w-full  B-ui-components_misc-input-ring "
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          style="position: relative; display: flex;"
        />
      </div>
    </div>
    <div
      class="flex flex-row justify-between px-2 my-1 text-xs "
    >
      <span
        class=""
      >
        Found 1 notes
      </span>
      <button
        class="focus:outline-none "
        data-bangle-editor-break="true"
        data-bangle-editor-pos="bottom"
        type="button"
      >
        Collapse all
      </button>
    </div>
    <div
      class="flex-grow"
      style="overflow-y: auto;"
    >
      <div
        class="B-ui-components_sidebar-row2 BU_hover B-search-notes_search-result-note-match pl-1 pr-3  select-none BU_active  "
        data-id="search-notes-result-0"
        role="button"
        style="cursor: pointer; display: flex; justify-content: space-between; user-select: none;"
      >
        <div
          class="flex flex-row"
        >
          <div
            class="B-ui-components_left-node"
          >
            <button
              class="focus:outline-none "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              <svg
                class=" "
                fill="currentColor"
                style="height: 16px; width: 16px;"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6.343 7.757L4.93 9.172 12 16.242l7.071-7.07-1.414-1.415L12 13.414 6.343 7.757z"
                />
              </svg>
            </button>
          </div>
          <span
            class="flex flex-col "
          >
            <span
              class="text-sm font-bold"
            >
              one
            </span>
            
          </span>
        </div>
        <div
          class="flex flex-row"
        >
          <span
            class="B-ui-components_right-node"
          >
            <button
              class="focus:outline-none text-xs font-semibold "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              1
            </button>
          </span>
          <span
            class="B-ui-components_right-hover-node"
          >
            <button
              class="focus:outline-none text-xs font-semibold "
              data-bangle-editor-break="true"
              data-bangle-editor-pos="bottom"
              type="button"
            >
              1
            </button>
          </span>
        </div>
      </div>
      <a
        class="rounded-sm block B-search-notes_search-result-text-match  mt-3 pb-4 border-b-1 border-colorNeutralBorder "
      >
        <div
          class="B-ui-components_sidebar-row2 BU_hover pl-3 "
          data-id="search-result-text-match-0"
          role="button"
          style="cursor: pointer; display: flex; justify-content: space-between; user-select: none;"
        >
          <div
            class="flex flex-row"
          >
            <div
              class="B-ui-components_left-node"
            />
            <span
              class=""
            >
              <span
                class="text-sm "
              >
                <div
                  class="B-search-notes_highlight-text-container"
                >
                  <span
                    class="B-search-notes_highlight-text text-sm"
                  >
                    hello
                  </span>
                  <span
                    class="B-search-notes_highlight-text text-sm"
                  >
                     world
                  </span>
                </div>
              </span>
            </span>
          </div>
          <div
            class="flex flex-row"
          >
            <span
              class="B-ui-components_right-node"
            />
            <span
              class="B-ui-components_right-hover-node"
            />
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
